# Daily Challenge: Classifying Handwritten Digits with CNNs

In [1]:
import tensorflow as tf
import tensorflow.keras as keras 

In [3]:
data = keras.datasets.mnist.load_data()
data[0][0].shape

(60000, 28, 28)

In [5]:
# Flatten the images from 28x28 to 784 pixels.
(x_train, y_train), (x_test, y_test) = data
x_train = x_train.reshape(-1, 784)
x_test = x_test.reshape(-1, 784)

In [6]:
x_train.shape, x_test.shape

((60000, 784), (10000, 784))

In [7]:
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

In [15]:
x_train.std()

0.30810776

In [16]:
# One-hot encode the target labels using keras.utils.to_categorical().
y_train = keras.utils.to_categorical(y_train, 10)
y_test = keras.utils.to_categorical(y_test, 10)

In [17]:
# Create a Sequential model.
model = keras.Sequential(
    [
        keras.layers.Input(shape=(784,)),
        keras.layers.Dense(128, activation="relu"),
        keras.layers.Dense(10, activation="softmax"),
    ]
)

In [18]:
model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"]
)
model.fit(x_train, y_train, epochs=5, batch_size=32, validation_split=0.2)
accuracy = model.evaluate(x_test, y_test)[1]

Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8722 - loss: 0.4668 - val_accuracy: 0.9517 - val_loss: 0.1638
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 7ms/step - accuracy: 0.9599 - loss: 0.1381 - val_accuracy: 0.9608 - val_loss: 0.1314
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9733 - loss: 0.0894 - val_accuracy: 0.9718 - val_loss: 0.0979
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9803 - loss: 0.0637 - val_accuracy: 0.9700 - val_loss: 0.1022
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 7ms/step - accuracy: 0.9832 - loss: 0.0531 - val_accuracy: 0.9733 - val_loss: 0.0917
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9712 - loss: 0.0936


In [19]:
accuracy

0.9753000140190125

In [21]:
# Reshape the input data to the shape expected by a Conv2D layer (e.g., (60000, 28, 28, 1)).
x_train = x_train.reshape(-1, 28, 28, 1)
x_test = x_test.reshape(-1, 28, 28, 1)

In [22]:
# cnn model
cnn = keras.Sequential(
    [
        keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        keras.layers.MaxPooling2D((2, 2)),
        keras.layers.Conv2D(64, (3, 3), activation='relu'),
        keras.layers.MaxPooling2D((2, 2)),
        keras.layers.Flatten(),
        keras.layers.Dense(64, activation='relu'),
        keras.layers.Dense(10, activation='softmax')
    ]
)
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
cnn.fit(x_train, y_train, epochs=5, batch_size=32, validation_split
=0.2)
cnn_accuracy = cnn.evaluate(x_test, y_test)[1]
cnn_accuracy

c:\Users\hp\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 29s 17ms/step - accuracy: 0.8826 - loss: 0.3696 - val_accuracy: 0.9808 - val_loss: 0.0634
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 38s 25ms/step - accuracy: 0.9828 - loss: 0.0540 - val_accuracy: 0.9855 - val_loss: 0.0504
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 49s 33ms/step - accuracy: 0.9883 - loss: 0.0348 - val_accuracy: 0.9888 - val_loss: 0.0382
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 40s 27ms/step - accuracy: 0.9910 - loss: 0.0286 - val_accuracy: 0.9858 - val_loss: 0.0527
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 40s 27ms/step - accuracy: 0.9936 - loss: 0.0198 - val_accuracy: 0.9882 - val_loss: 0.0435
313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step - accuracy: 0.9852 - loss: 0.0489


0.9886999726295471

as we can see the accuracy of the CNN model (99%) is bigger than the accuracy of the fully connected sequential model (97%), so our CNN model is more accurate as it capture more complex pattern and details.